snapcraft/plugins/dotnet.py

```diff
@@ -0,0 +1,200 @@
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - dotnet-runtime:
+      (string)
+      The runtime to embed in the resulting snap.
+"""
+import os
+import shutil
+import fnmatch
+
+import snapcraft
+from snapcraft import sources
+
+
+_RUNTIME_DEFAULT = '2.0.0'
+_SDK_DEFAULT = '2.0.0'
+# TODO extend for more than xenial
+_SDKS_AMD64 = {
+    '2.0.0': dict(url_path='http://dotnetcli.blob.core.windows.net/dotnet/'
+                           'Sdk/2.0.0/dotnet-sdk-2.0.0-linux-x64.tar.gz',
+                  checksum='sha256/6059a6f72fb7aa6205ef4b52583e9c041f'
+                           'd128e768870a0fc4a33ed84c98ca6b')
+              }
+# TODO extend for other architectures
+_SDK_DICT_FOR_ARCH = {
+    'amd64': _SDKS_AMD64,
+}
+
+
+class DotNetPlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+
+        schema['properties']['dotnet-runtime'] = {
+            'type': 'string',
+            'default': _RUNTIME_DEFAULT,
+        }
+
+        if 'required' in schema:
+            del schema['required']
+
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['dotnet-runtime']
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['dotnet-runtime', 'build-attributes']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._dotnet_dir = os.path.join(self.partdir, 'dotnet')
+        self._dotnet_sdk_dir = os.path.join(self._dotnet_dir, 'sdk')
+
+        # These are required for the runtime
+        if 'no-runtime' in options.build_attributes:
+            self._runtime = _DummyTar()
+        else:
+            self.stage_packages.extend([
+                'libcurl3',
+                'libcurl3-gnutls',
+                'libicu55',
+                'liblttng-ust0',
+                'libunwind8',
+                'lldb',
+                'libssl1.0.0',
+                'libgssapi-krb5-2',
+                'libc6',
+                'zlib1g',
+                'libgcc1'
+            ])
+
+        self._sdk = self._get_sdk()
+        self._dotnet_cmd = os.path.join(self._dotnet_sdk_dir, 'dotnet')
+
+    def _get_sdk(self):
+        try:
+            sdk_arch = _SDK_DICT_FOR_ARCH[self.project.deb_arch]
+        except KeyError as missing_arch:
+            raise NotImplementedError(
+                'This plugin does not support architecture '
+                '{}'.format(missing_arch))
+        try:
+            # TODO support more SDK releases
+            sdk_version = sdk_arch['2.0.0']
+        except KeyError as missing_version:
```

Comment:
This is an exception that will never be reached, because the '2.0.0' is hardcoded.